src/macros-by-example.md

```diff
@@ -24,149 +24,404 @@
+an  `extern crate` declaration. Macros imported this way are imported textually
+only. Optionally, a list of macros to import can be specified; this is not
+supported when `#[macro_use]` is applied to a module.
+
+```ignore
+#[macro_use(lazy_static)] // Or #[macro_use] to import all macros.
+extern crate lazy_static;
+
+lazy_static!{};
+// self::lazy_static!{} // Error: lazy_static is not defined inself
+```
+
+Macros to be imported with `#[macro_use]` must be exported with
+`#[macro_export]`, which is described below.
+
+### Path-Based Scope
+
+By default, a macro has no path-based scope. However, if it has the
+`#[macro_export]` attribute, then it is declared in the crate root scope and can
+be referred to normally as such:
+
+```
+self::m!();
+m!(); // OK: Path-based lookup finds m in the current module.
+
+mod inner {
+    super::m!();
+    crate::m!();
+}
+
+mod mac {
+    #[macro_export]
+    macro_rules! m {
+        () => {};
+    }
+}
+```
+
+Macros labeled with `#[macro_export]` are always `pub` and can be referred to
+by other crates, either by path or by `#[macro_use]` as described above.
+
+## Hygiene 
+
+By default, all identifiers referred to in a macro are expanded as-is, and are
+looked up at the macro's invocation site. This can lead to issues if a macro
+refers to an item or macro which isn't in scope at the invocation site. In order
+to alleviate this, the `$crate` metavariable can be used at the start of a path
+in order to force lookup to occur inside the crate defining the macro.
+
+```ignore
+//// Definitions in the `helper_macro` crate.
+#[macro_export]
+macro_rules! helped {
+    // () => { helper!() } // This might lead to an error due to 'helper' not being in scope.
+    () => { $crate::helper!() }
+}
+
+#[macro_export]
+macro_rules! helper {
+    () => { () }
+}
+
+//// Usage in another crate.
+// Note that `helper_macro::helper` is not imported!
+use helper_macro::helped;
+
+fn unit() {
+    helped!();
+}
+```
+
+Note that, because `$crate` refers to the current crate, it must be used with a
+fully qualified module path when referring to non-macro items:
+
+```
+pub mod inner {
+    #[macro_export]
+    macro_rules! call_foo {
+        () => { $crate::inner::foo() };
+    }
+
+    pub fn foo() {}
+}
+```
+
+Additionally, even though `$crate` allows a macro to refer to items within its
+own crate when expanding, its use has no effect on visibility. An item or macro
+referred to must still be visible from the invocation site. In the following
+example, any attempt to invoke `call_foo!()` from outside its crate will fail
+because `foo()` is not public.
+
+```
+#[macro_export]
+macro_rules! call_foo {
+    () => { $crate::foo() };
+}
+
+fn foo() {}
+```
+
+> **Version & Edition Differences**: Prior to Rust 1.30, `$crate` and
+> `local_inner_macros` (below) were unsupported. They were added in
+> preparation for Rust 2018's path-based imports of macros (described above), in
```

Comment:
Path-based imports are not specific to Rust 2018, it's just "before Rust 1.30" and "after Rust 1.30".